library/core/src/task/wake.rs

```diff
@@ -541,6 +541,16 @@ impl Waker {
     ///
     /// If an owned `Waker` is needed, `clone()` this one.
     ///
+    /// # Warning
+    ///
+    /// Do not use this function as a way to be able to call `poll` methods,
+    /// in production code,
+    /// from within a non-`async`, non-`poll`, function.
+    ///
+    /// Using a no-op waker for that purpose would cause wakeups to be lost:
```

Comment:
This text is, in my opinion, either incorrect or not specific enough about the situation it is warning against.

`Waker::noop()` is a perfectly good way to _implement executing a task_ in the case where it is an acceptable performance tradeoff, or suitable for the application, to poll the task every time through some sort of loop (or to poll the task exactly once, as in [`now_or_never()`](https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.now_or_never)). No wakeups are “lost” because the task is unconditionally awake. This is, to my mind, the main use case of `Waker::noop()`: to implement async execution trivially when it isn't worth the complexity of implementing it better.

It can only cause lost wakeups if you *mix* polling the task this way with polling the task another way.  I get the impression from https://github.com/rust-lang/rust/issues/98286#issuecomment-2243447136 that your use case is perhaps doing something like that, but the text in this change is talking about non-async functions only, not mixing polling from a non-async function and polling from an executor.

Could you rephrase this to be more specific about the situation you are warning against?

Reply:
Thanks.  I'll see waht I can do.

I came here to do this because in conversations with colleagues, some of whom are fairly new to Rust, especially async Rust, hadn't understood the implications.

I can see that there can be situations where wakers aren't important, particularly in an *executor*.  I think using a noop waker in a future implementation, without juggling caller(s)' waker(s), is probably a bug.  Am I wrong about that?

Reply:
>  I think using a noop waker in a future implementation, without juggling caller(s)' waker(s), is probably a bug. Am I wrong about that?

Yes, if you are writing a future combinator (that is, `impl Future for...` that delegates to other `Future`s), then you should usually _either_ pass on the provided context's waker, or create your own delegating non-noop wakers (if you want to track which of several child futures woke).

But the text doesn't say “when you are implementing `Future::poll`”. It says “from within a non-`async`, non-`poll`, function”, which *rules out* that case. So, I am confused.

Reply:
> since it involves shared ownership of the task, or using `noop()` within a future combinator

Right.  That is the scenario I'm trying to warn about, particularly using noop within a futures combinator.

I have tried to improve the text to be more specific.  What do you think now?  (I've avoided the term "future combinator" since it seems rather abstruse, and the reader who is about to make the mistake I'm warning about may not realise that that's what they're writing.)

Reply:
> What do you think now?

That’s specific enough that it no longer has the “vague scary warning” problem, yes. Further comments:

* I don't think the words “runtime-agnostic” are adding anything here; if the future has knowledge about the runtime then that potentially overrides *everything* about the conventional `Future`/`Waker` protocol, so it doesn't need to be called out here.

* It’s still possible to use a noop waker correctly in this scenario. In particular, the usual time a wrapper future is likely to be without `Context` is at construction, and futures are always polled once to start. So, if `Foo` wraps `Bar` and polls `Bar` with a noop waker when `Foo` is constructed, then that's fine as long as `Foo` also polls `Bar` with the propagated waker when `Foo` is polled for the first time. I don't have an idea for how to describe this sort of scenario concisely, though.

> I've avoided the term "future combinator" since it seems rather abstruse

Yes, and pedantically, you can have a future that wraps other futures without it being purely a combinator, so this is more correct.

Reply:
I understand what, @ijackson, you're trying to do here, and I appreciate the intent.  And so it kind of saddens me to be so negative on this change, but...

> When implementing a runtime-agnostic future which wraps other future(s), do not use this function as a way to be able to call `poll` methods when you don't have your caller's `Context` available. Using a no-op waker for that purpose would cause wakeups to be lost: the inner future would store only the no-op waker, replacing the proper waker for the waiting task.

In addition to the good points that @kpreid [raised](https://github.com/rust-lang/rust/pull/128064/files#r1700481858), when I read this, I think to myself, "anyone who is able to parse and interpret those two sentences already knows the limitations without needing to be told."

That's what I think the nature of the problem is here.  It seems harder to write an accessible and correct warning than it is to just write an accessible statement of what this does do.

Reply:
Actually, how about something like this? Expanding the current paragraph that says “This is mostly useful for writing tests…”, we could write:

> This is mostly useful for writing tests that need a `Context` to poll some futures, but are not expecting those futures to wake the waker or do not need to do anything specific if it happens. More generally, using `Waker::noop()` to poll a future **discards the notification of when the future should be polled again**, and therefore, it should only be used when such a notification will not be needed to make progress.

I think this would fairly well get across some useful points:

* What you are doing here is discarding a message. (People understand that this can mean that progress stops because the message is not delivered where it is needed.)
* It is reasonable to discard a message specifically when nothing needs to be done about it.

without getting further into the details of precisely classifying what situations this applies to.

Reply:
I like your suggested text.  Thank you!  I will adopt it into the MR.